### PR TITLE
fix: skip `head merge commits` on pull requests

### PR DESCRIPTION
**Pull request by grouville**
Fixes https://linear.app/dagger/issue/DEV-3086/pull-user-commit-message-in-a-context-pr

When collecting labels on the engine in a VCS context, sometimes, the commit head is an appended commit (from that VCS) [PR / MR]. This fix skips the merge request commit when it is the head.

Repro/test: https://gist.github.com/grouville/6cf442d86c41c0746bbef1812160e7c0

### 12/13 Advancements

> Realization: only GitHub puts the PR in a detached HEAD with only the merge commit. GitLab keeps the git history of the branch being tested

After reviews + some testing, a few changes have been implemented:
1. Make sure that the main branch doesn't skip the merge commits. Only `pull_requests` shall be be impacted
2. Due to the realization, we need to fetch the branch and get the commits from there. However, this is an operation requiring authorization. To circumvent this, we rely on the fact that GitHub runners automatically add the `GITHUB_TOKEN` secrets to git CLI commands

### 12/15 Advancements
Previous implementation was introducing a bug by printing errors to stdout, which polluted the `dagger listen` connection string.

The error came from the fact that on Dagger, the PRs come from a fork and not the origin. However, after spending some time reverse engineering Github Actions' runner context, I did not find a 100% solution to retrieve the origin of a fork. This is a best try approach, where we build the fork's remote of a repo from available information. If we do not manage to do it, then we fallback on the current logic

### CODE REVIEW

File: core/pipeline/label.go [hunk 1]
```diff
@@ -139,12 +140,29 @@ func LoadGitLabels(workdir string) ([]Label, error) {
 		return nil, err
 	}
 
+	// Check if the commit is a merge commit (GitHub context)
+	if commit.NumParents() > 1 {
```
Comment (jedevc) [inline core/pipeline/label.go:144]: So wouldn't this also trigger on main, when doing merge commits (instead of rebase/squash)?

In that case, the message *should* be the merge commit, right? (since it was the last commit on that branch)

Can we use something else to identify this magic commit? I can't find any docs on this, but I wonder if the commit message has more details, or whether the commit author is an automated github thing?

File: core/pipeline/label.go [hunk 2]
```diff
@@ -139,12 +140,31 @@ func LoadGitLabels(workdir string) ([]Label, error) {
 		return nil, err
 	}
 
+	// Check if the commit is a merge commit in the context of PR / MR
+	if (os.Getenv("GITHUB_EVENT_NAME") == "pull_request" ||
+		os.Getenv("CI_PIPELINE_SOURCE") == "merge_request_event") &&
+		commit.NumParents() > 1 {
```
Comment (marcosnils) [inline core/pipeline/label.go:146]: is it possible for a MR to **not** trigger this condition? Should we log something if that's not the case?

File: core/pipeline/label.go [hunk 3]
```diff
@@ -139,12 +140,31 @@ func LoadGitLabels(workdir string) ([]Label, error) {
...
+			return nil, err
+		}
+
+		// Skip the first parent and use the second parent as the previous commit
+		parents, err = parents.Parents().Next()
+		if err != nil {
+			return nil, err
+		}
+
+		commit = parents
```
Comment (marcosnils) [inline core/pipeline/label.go:159]: nit: call `commit.Parent(n)` to simplify?

observation: I'm surprised that it's the 2nd parent that has the actual commit we're looking for. Github and Gitlab create two commits per PR?

File: core/pipeline/label.go (hunk 1 above)
Comment (marcosnils) [reply core/pipeline/label.go]: > So wouldn't this also trigger on main, when doing merge commits (instead of rebase/squash)?

I don't think so since here https://github.com/dagger/dagger/pull/6253/files#diff-8b206da82ce6c3970baf564e295175e55119318f3a46ca53d8b90dde0b266b34R144-R145 we're specifically checking if this is a Github PR or Gitlab MR.

File: core/pipeline/label.go [hunk 4]
```diff
@@ -168,18 +188,27 @@ func LoadGitLabels(workdir string) ([]Label, error) {
...
+				labels = append(labels, Label{
+					Name:  "dagger.io/git.branch",
+					Value: ref.Name().Short(),
+				})
+			}
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
```
Comment (marcosnils) [inline core/pipeline/label.go:211]: just to be sure, is this change required because before we were incorrectly collecting the tag and the branch in PR's?

File: core/pipeline/label.go (hunk 2 above)
Comment (grouville) [reply core/pipeline/label.go:146]: if thats not the case, it will have the same behavior as today, the git title will be: Merge <hash> to <hash>, so need to log I thing, we'll see it easily

File: core/pipeline/label.go (hunk 4 above)
Comment (grouville) [reply core/pipeline/label.go:211]: it could work as in the past, however, the tag is not applied to the added merge commit from GitHub but to a previous commit, so we would miss it

Comment (grouville) [reply core/pipeline/label.go:223]: > just to be sure, is this change required because before we were incorrectly collecting the tag and the branch in PR's?

You are right. In the context of a PR, we would 100 % miss the tag.

**Comment by grouville** [comment]
@marcosnils the failing test does not seem to come from me. When running this test locally, it seems to work

File: core/pipeline/label.go [hunk 5]
```diff
@@ -97,6 +100,78 @@ func LoadVCSLabels(workdir string) []Label {
...
+	username := os.Getenv("GITHUB_ACTOR")
+
+	// Get the repository name (owner/repo)
+	repository := os.Getenv("GITHUB_REPOSITORY")
+	parts := strings.Split(repository, "/")
+	if len(parts) < 2 {
+		return nil, fmt.Errorf("invalid repository format: %s", repository)
+	}
+
+	forkURL := "https://github.com/" + username + "/" + parts[1]
```
Comment (jedevc) [inline core/pipeline/label.go:145]: Agh, I don't love this, it feels very potentially hacky (one note, we can also use `GITHUB_SERVER_URL` here, to work with private GitHub instances).

That said, I don't have better suggestions if GitHub isn't doing anything to properly label their merge commits as auto-generated. I'd be happy to get this in sooner rather than later if this resolves 90% of cases (which it seems it probably will) - but think we'll have to revisit at some point since it's not just GitHub who does this behavior.

Comment (grouville) [reply core/pipeline/label.go:145]: Great comment, implemented

**Review by jedevc — APPROVED** [review]
